eng/ILLink.Substitutions.Resources.template

```diff
@@ -0,0 +1,9 @@
+<linker>
+  <assembly fullname="{AssemblyName}">
+    <!-- System.SR.UsingResourceKeys removes resource strings and instead uses the resource key as the exception message -->
+    <resource name="{StringResourcesName}.resources" action="remove" feature="System.SR.UsingResourceKeys" featurevalue="true" />
```

Comment:
"SR" sounds pretty cryptic. Is `System.SR.UsingResourceKeys` a good public facing name? Should it rather be `System.Resources. ...` ?

Reply:
SR in general is internal and not public. Are you talking about the name used here in these targets? does it matter much?

Reply:
it will be public in the sense that people can pass that in as extra args to the trimmer if they want to manually define it that way. Most people though will be instead using a property that will be added to the SDK that will control this feature switch, with a name like:
```xml
<PropertyGroup>
  <TrimResources>true</TrimResources>
</PropertyGroup>
```
or something like that. All that said though, I do agree with @jkotas point plus we are also using the same name witht he AppContext switch which will be definitely "more public" as people will be setting this on their runtimeconfig file. I will change the name.

Reply:
One concern I have with `System.Resources.UseResourceKeys` is that it sounds like it affects _all_ resources, not just the exception messages coming from the runtime.

Is there a word/term/name we can use to make it specific to the runtime resources?

* `System.Resources.UseSystemResourceKeys`
* `System.Resources.UseBclResourceKeys`
* `System.Resources.UseSRResourceKeys`

Reply:
> AppContext switch which will be definitely "more public" as people will be setting this on their runtimeconfig file. I will change the name

while it's possible to set this directly in `.runtimeconfig.json` I don't expect many people actually doing it. We expect the vast majority to use MSBuild property to drive this.

So we need a really good "customer facing" name for the MSBuild property.
And we need a good "technical" name for the runtime option/AppContext switch.

What the options does:
When enabled, all string resources in our framework assemblies will be removed and replaced with the resource keys instead. Framework only uses string resources for exception messages (I'm not aware of any other usage), so the effect of this change is for example (aside from saving quite a bit of size):
```
System.ArgumentNullException: Value cannot be null. (Parameter 'args')
```
turns into 
```
System.ArgumentNullException: ArgumentNull_Generic Arg_ParamName_Name, args
```

Customer facing name should reflect the behavioral change - so maybe something along the lines of `UseResourceKeysForExceptionMessages` (this is a bad name, but the intent is there).

Technical name should be something closer to the code base, so maybe `System.Resources.UseStringSystemResourceKeys`...

Adding @samsp-msft who might have better ides about naming.

Reply:
>  Framework only uses string resources for exception messages (I'm not aware of any other usage)

Here are examples of resource uses for strings that are not exception messages:
- https://github.com/dotnet/runtime/search?q=Globalization_cp_&unscoped_q=Globalization_cp_
- https://github.com/dotnet/runtime/search?q=Word_At&unscoped_q=Word_At

Reply:
Thanks @jkotas! I guess it's OK to remove resources for those as well... it just makes the naming more complicated...

Reply:
> Here are examples of resource uses for strings that are not exception messages:

I assume for .NET Native, which used this same switch, those places were broken as well? For example in a StackTrace.ToString() you would get `Word_At` instead of `at`?

Reply:
That's what `defaultString` argument of `GetResourceString` method was designed for. IIRC, the most egregious cases were fixed by passing in a sensible default string into this API in .NET Native. I do not remember whether `Word_At` got this treatment.

Reply:
> That's what defaultString argument of GetResourceString method was designed for.

I see. My assumption was that `defaultString` was hard-coded in C# during Debug builds, so we had nice exception messages when we were debugging. But I guess it would make sense to use it elsewhere as well.